### ISSUE TITLE
Fixes to daily subsurface routing 

### ISSUE THREAD

**Issue by laurencelin** [RHESSys/RHESSys#84]
In trying to debug large unsat_drain and large sat_def daily simulation in rhessys 5.20, found the following two problems in compute_subsurface_routing.c
1) these calculation are duplicated in every time step when previously they were only calculated before the time loop
https://github.com/RHESSys/RHESSys/blob/master/rhessys/hydro/compute_subsurface_routing.c#L214
https://github.com/RHESSys/RHESSys/blob/master/rhessys/hydro/compute_subsurface_routing.c#L219

and 
2) This finalization streamflow and sat deficit are now preforming every time step when previously it was performed only at the last time step.
https://github.com/RHESSys/RHESSys/blob/master/rhessys/hydro/compute_subsurface_routing.c#L368

This issue will be addressed i the subsequent branch and pull-quest. 
In our testing, this is a major bug in this version. It should be addressed in the master branch as well as merge to the develop.

**Comment by tungmilan** [comment]
@laurencelin Isn't it necessary to update patch[0].preday_sat_deficit and patch[0].sat_deficit after every timestep k? From my understanding,  those lines should be interpreted as "patch pre timestep sat deficit". 

patch[0].preday_sat_deficit and patch[0].sat_deficit values change within k-loop here
https://github.com/RHESSys/RHESSys/blob/master/rhessys/hydro/compute_subsurface_routing.c#L267

**Comment by laurencelin** [comment]
patch[0].sat_deficit and patch[0].sat_deficit_z values change within k-loop

But patch[0].preday_sat_deficit and patch[0].preday_sat_deficit_z are not used for every k-time-step calculations (in RHESSys 5.18.r2). patch[0].preday_sat_deficit_z is used for field_capacity calculation in the "n_time_step-1" th step within the k-loop (according to RHESSys 5.18.r2 subsurface_routing).  

Note that there is a modification (https://github.com/RHESSys/RHESSys/blob/master/rhessys/hydro/compute_subsurface_routing.c#L368) in RHESSys 5.20, which causes highly unstable and extremely large values to daily unsat_drain and daily sat_def_z.  We did not have this problem in 5.18.  Therefore I try to roll back the 5.20-modification in subsurface_routing to 5.18.

**Comment by tungmilan** [comment]
Thanks @laurencelin . Tracking back, @xiaoli-chen made those changes to the code. Would you mind explaining the reasoning behind them @xiaoli-chen ? Thanks!

https://github.com/RHESSys/RHESSys/commit/a107b761a3e112eed3f24e03d52eab0a5bfe6508#diff-d2abc97ff4e44a705ad0a98517300bf7R212

https://github.com/RHESSys/RHESSys/commit/9e2faecf7035ba8609b3f4fb499ba4c6d7e8f493#diff-d2abc97ff4e44a705ad0a98517300bf7R359

**Comment by selimnairb** [comment]
@xiaoli-chen As per @tungmilan question, you can also refer to @laurencelin pull request #85.

**Comment by xiaoli-chen** [comment]
@laurencelin @tungmilan Thanks for tracking this issue.
I think both issues may trace back to the modification for hourly simulation.

**Why finalization streamflow and sat deficit are now preforming every time step?**
    Daily finalization streamflow is not calculated at every time step. It is still finalized at the end of the day:
https://github.com/RHESSys/RHESSys/commit/a107b761a3e112eed3f24e03d52eab0a5bfe6508#diff-d2abc97ff4e44a705ad0a98517300bf7R980

    The sat deficit, infiltration and excess water to detention store is set to calculated every time step. 
    Then why change the if (k == n_timestep-1) to if (k> =0) in 5.20 daily calculation compute_subsurface_routing.c? It is because I try to make the daily result consistent with the hourly calculation.

    In daily mode, 5.18 compute_subsurface_routing.c has done the calculation in following procedure:
   a) from hour 1 to hour 24, update_drainage_road/stream/land is called to calculate the throughflow, then update moisture.
   b) at hour 24, following calculation is done: calculate excess water to detention store -> overland flow routing ->  field capacity -> infiltration -> sat_deficit -> field capacity -> finalize daily streamflow

If you output the hourly sat_deficit from 5.18 compute_subsurface_routing.c, you will find the output is reasonable from hour 1 to hour 23, but inconsistent between hour 23 and hour 24. This is because in 5.18 compute_subsurface_routing.c, RHESSys has done too many thing only at hour = 24. 

   When I modify RHESSys to output hourly result, obviously I don't want to see the inconsistency between hour 23 and hour 24. To make the the hourly output consistent, I update 5.20 compute_subsurface_routing_hourly.c in following way:
    a) from hour 1 to hour 24, update_drainage_road/stream/land is called to calculate the throughflow, then calculate excess water to detention store -> overland flow routing -> field capacity -> infiltration -> sat deficit -> field capacity -> hourly streamflow is calculated at each hour. 
    b) at hour 24, daily streamflow is finalized at the last hour of the day
https://github.com/RHESSys/RHESSys/blob/master/rhessys/hydro/compute_subsurface_routing_hourly.c#L374
    Above is the hourly calculation in 5.20. 
    
   To make the daily calculation consistent with the hourly routing for subsurface routing, I change the compute_subsurface_routing.c in 5.20 to the following procedure:
    a) from hour 1 to hour 24, update_drainage_road/stream/land is called to calculate the throughflow, then calculate excess water to detention store -> overland flow routing -> field capacity -> infiltration -> sat deficit -> field capacity -> hourly streamflow is calculated at each hour. 
    b) at hour 24, daily streamflow is finalized at the last hour of the day

    It is consistent with compute_subsurface_routing_hourly.c in 5.20. 

    That is why I change the if statement from (k==n_timesteps) to (k>=0). 

    This modification may bring bugs and weird results. But i dont' think the (k>=0) is the real problem. It is the long chain of calculation: calculate excess water to detention store -> overland flow routing -> field capacity -> infiltration -> sat deficit -> field capacity. When calculate once a day (previous setting in 5.18), it is OK. However, when calculating 24 times a day (in 5.20), any tiny bugs in this chains of calculation will be enlarged. 
    
    Sorry for the late and long reply. 

Xiaoli

**Comment by laurencelin** [comment]
@Xiaoli thanks for the update. I basically agree to you about that n_time_step -1.  But this subsurface_routing is a daily routine, right? In the files, it seems you have made a new hour routine subsurface_routing.hourly.c.   In this case, could we keep the daily subsurface routing as it was? I would like see its improvement. Changing to k>=0 may be the first step but it's not completed. As a result unsat drain and sat def z became unstable and large, as well as shifting the entire hillslope hydrology behavior: water table stays near to the deepest soil depth most of the time while unsat drain is always extremely large in our simulation in Coweeta. 

Laurence

**Comment by xiaoli-chen** [comment]
@laurencelin Sure, you can create another branch and try changing the if statement to k==n_time_step-1 and see whether it helps.

**Comment by tungmilan** [comment]
Thanks @xiaoli-chen for the clarification! So that explains these lines
https://github.com/RHESSys/RHESSys/commit/a107b761a3e112eed3f24e03d52eab0a5bfe6508#diff-d2abc97ff4e44a705ad0a98517300bf7R212
You wanted to update sat_deficit & sat_deficit_z after every hour. Am I correct?

**Comment by xiaoli-chen** [comment]
@tungmilan Yes, you are correct. And the preday_sat_deficit are actually now prehour_sat_deficit.
